superset-frontend/src/views/CRUD/data/database/DatabaseModal/DatabaseConnectionForm.tsx

```diff
@@ -34,27 +45,170 @@ export const FormFieldOrder = [
+  );
+  return (
+    <CredentialInfoForm>
+      {!isEditMode && (
+        <>
+          <FormLabel required>
+            `${t('How do you want to enter service account credentials?')}`
+          </FormLabel>
+          <Select
+            defaultValue={uploadOption}
+            style={{ width: '100%' }}
+            onChange={option => setUploadOption(option)}
+          >
+            <Select.Option value={CredentialInfoOptions.jsonUpload}>
+              `${t('Upload JSON file')}`
+            </Select.Option>
+
+            <Select.Option value={CredentialInfoOptions.copyPaste}>
+              `${t('Copy and Paste JSON credentials')}`
+            </Select.Option>
+          </Select>
+        </>
+      )}
+      {uploadOption === CredentialInfoOptions.copyPaste ||
+      isEditMode ||
+      editNewDb ? (
+        <div className="input-container">
+          <FormLabel required>`${t('Service Account')}`</FormLabel>
+          <textarea
+            className="input-form"
+            name="credentials_info"
+            value={db?.parameters?.credentials_info}
+            onChange={changeMethods.onParametersChange}
+            placeholder={JSON.stringify(
+              {
+                credentials_info: '<contents of credentials JSON file>',
+              },
+              null,
+              '  ',
+            )}
+          />
+          <span className="label-paste">
+            `${t('Copy and paste the entire service account .json file here')}`
+          </span>
+        </div>
+      ) : (
+        <div
+          className="input-container"
+          css={(theme: SupersetTheme) => infoTooltip(theme)}
+        >
+          <div css={{ display: 'flex', alignItems: 'center' }}>
+            <FormLabel required>`${t('Upload Credentials')}`</FormLabel>
+            <InfoTooltip
+              tooltip={t(
+                'Use the JSON file you automatically downloaded when creating your service account in Google BigQuery.',
+              )}
+              viewBox="0 0 24 24"
+            />
+          </div>
+
+          {!fileToUpload && (
+            <Button
+              className="input-upload-btn"
+              onClick={() => document?.getElementById('selectedFile')?.click()}
+            >
+              `${t('Choose File')}`
+            </Button>
+          )}
+          {fileToUpload && (
+            <div className="input-upload-current">
+              {fileToUpload}
+              <DeleteFilled
+                onClick={() => {
+                  setFileToUpload(null);
+                  changeMethods.onParametersChange({
+                    target: {
+                      name: 'credentials_info',
+                      value: '',
+                    },
+                  });
+                }}
+              />
+            </div>
+          )}
+
+          <input
+            id="selectedFile"
+            className="input-upload"
+            type="file"
+            onChange={async event => {
```

Comment:
small nit, but this is a fairly long function in this event handler. Not necessarily required now, but I think it would read better if we pulled this out into a named function instead.